### ISSUE TITLE
Create a provenance table for function results tracking

### ISSUE THREAD

**Issue by ladoramkershner** [USGS-Astrogeology/autocnet#582]
Every apply jobs produces a result, that equates to the status message(s) given out by the applied function. We currently attach the result to a log status and the entire log message (which also includes function parameterize) gets printed out to the log file. The current information getting tracked is fairly thorough, but the information would be much more usable if it was put into a 'provenance' database table as opposed to a log file. 

Finding the relevant information currently requires grepping hundreds of thousands of log files which takes ~20 min to run. Then, if you have one 'space' out of place, your grep will return nothing.. and you have to start all over. It makes it really difficult to debug.

The major issue to solve with adding this, is that we cannot guarantee that every workflow will use the same processes. So the table would need to be dynamic in terms of number/naming of columns. We may be able to create a bare bones provenance table with foreign key information to imageid, pointid, measureid (all of which can be null) then add a column based on the chooser keyword passed into ncg.apply(). I am not sure if the add a column could be added as a function under the orm Provenance model, but that function can get called at the top of every ncg.apply call.

**Comment by Kelvinrr** [comment]
Do you have an idea of what the schema for such a table would look like? Basically, specifically what you would want to store. 

Postgres has powerful json tools so we might be able to leverage that (e.g. store kwargs). From what I understand is, whenever we apply a function to a measure, point or image, we store the ID, the function name, function parameters, and some kind result? The former two are easy but the last one is kinda tricky.

**Comment by jessemapel** [comment]
Maybe we can record the results by tracking changes to the ORM with something like this https://stackoverflow.com/questions/29921260/tracking-model-changes-in-sqlalchemy

**Comment by acpaquette** [comment]
@Kelvinrr I agree with storing the JSON but from experience with PDS and large JSON stores in significantly large databases, it can be difficult to run queries against those JSON stores. I have only experimented with this briefly, so I could be wrong, but the majority of queryable aspects should probably be stored in columns in basic types like we have been doing for the majority of autocnet databases. From your comment something like storing the "...measure, point or image, ... the ID, the function name..." as common data types would make things much faster to query, then hold the function parameters as JSON on the record in the table. 

I'm sure you've given this some thought, I just figured I would chime in and mention that directly querying JSON can be seriously slow and painful.

**Comment by Kelvinrr** [comment]
From looking at [some summaries of how to best handle table history for PSQL tables](https://dba.stackexchange.com/questions/233735/track-all-modifications-to-a-postgresql-table) it seems like history table with 1-m mapping with the candidate table might be the cleanest way to do this? So like, images_history, point_history, measure_history tables. And keeping them updated using the method Jessie suggested of having the triggers on the Model and not as SQL. 

A history tables schema I think would work: 

* id
* timestamp
* fk (basically FK to host table's ID, like pointId, imageid, or measure id)
* functionName 
* args (kwargs used stored as jsonb)
* before (jsonb of the row values before change)
* after (jsonb of the row values after change)

the bug downside is these tables will grow very large very quick as jobs are run, but I think there is no real way of avoiding that.

**Comment by jessemapel** [comment]
One last thing to consider is how to handle a failure or if a function runs but makes no modifications to a model object. Maybe that's just a separate "failures" table.
